---
title: "GWAS_LiftOver"
author: "Nick Burns"
date: "19 September 2016"
output: html_document
---

## GWAS LiftOver  

We need to try to standardise our GWAS summary data for GenomeDB. This involves making sure all data is on hg19, and we need to also lift over the rsids. We cover both of these below, following this guide: http://genome.sph.umich.edu/wiki/LiftOver  

### hg18 -> hg19  

We will begin by lifting over the coordinates using the web-based liftOver tool: http://genome.ucsc.edu/cgi-bin/hgLiftOver  

```{r}
library(data.table)
data_dir <- "/mnt/DataDrive/GWASDatasets/"

write.bed <- function (x, filename) {
    write.table(x, filename, sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}

diagram <- fread(paste0(data_dir, "diagram.mega-meta.txt"), stringsAsFactors = FALSE)

diagram[, CHR := paste0('chr', CHR, sep = "")]
diagram[, Start := as.integer(POS)]
diagram[, End := as.integer(Start + 1)]

write.bed(diagram[, .(CHR, Start, End, SNP)],
          paste0(data_dir, "diagram_pre_lift_over.tsv", sep = ""))
```

Then run the following command line tool to lift over the coords:

```
./liftOver <chain_file.gz> <input_file> <output_file> <failed_log>
```


Let's look at the lines that failed:

```{r}
failed <- read.table("/mnt/DataDrive/GWASDatasets/Curated/LiftOver/failed_covnersion.bed")
head(failed)

failed <- data.table(failed)
colnames(failed) <- c("CHR", "Start", "End", "SNP")

failed <- merge(diagram, failed, by = c("CHR", "Start", "End"))
summary(failed$P)
```

Good news, is that we didn't lose any peaks which reach genome-wide significance.


### RSID liftOver  

