---
title: "GWAS_Build37_Updates"
author: "Nick Burns"
date: "6 September 2016"
output: html_document
---

## GWAS Build 37 liftover

Many of these GWAS results are pre-build 37. Here, we will update each result set to build 37 and then update GenomeDB where possible. The kottgen dataset is fine, I need to update diagram, GIANT and locke.

```{r}
setwd("G:/Datasets/GenomeDBData/GWASDatasets/")
library(data.table)
library(glida)

gwas_files <- list(
  diagram = "diagram.mega-meta.csv",
  giant = "GIANT_footer_removed.csv",
  locke = "LockeSummary.csv"
)
```

For each of these files, we will update the SNP positions (where possible) from dbSNP142 (using the GLIDA package). Where SNPs cannot be matched to this db, we will simply leave them untouched in the database. Will keep this in mind as we continue to explore the data.

```{r}
for  (gwas in gwas_files) {
    print(sprintf("GWAS FILE:  %s", gwas))
    tmp <- fread(gwas)
    print(head(tmp))
    
    snpcol <- colnames(tmp)[1]    # SNPs are always the first column
    setkeyv(tmp, snpcol)
    
    # update snps in groups of 100000
    idx <- 1:nrow(tmp) %/% 100000
    for (i in unique(idx)) {
       
       snps <- unlist(tmp[idx == i, snpcol, with = F])
       
       updated_snps <- data.table(glida::queryUCSC(glida::updatePositions(snps)))
       setkeyv(updated_snps, "SNP")
       
       tmp[updated_snps, POS := updated_snps$POS]
    }
    
    write.csv(tmp, paste0(gwas_file, ".build_37.csv", sep = ""), row.names = F, quote = F)
}

```